File: server/src/main/java/bgu/spl/net/impl/stomp/Frame.java
package bgu.spl.net.impl.stomp;

import java.util.HashMap;
import java.util.Map;


public class Frame {

    private String command;
    private Map<String, String> headers;
    private String body;

    public Frame(String command, Map<String, String> headers, String body) {
        this.command = command;
        this.headers = headers != null ? new HashMap<>(headers) : new HashMap<>();
        this.body = body;
    }

    public Frame() {
        this.headers = new HashMap<>();
    }

    public String getCommand() {
        return command;
    }

    public void setCommand(String command) {
        this.command = command;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public void addHeader(String key, String value) {
        headers.put(key, value);
    }

    public void removeHeader(String key) {
        headers.remove(key);
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    @Override
    public String toString() {
        StringBuilder frame = new StringBuilder(command + "\n");
        headers.forEach((key, value) -> frame.append(key).append(":").append(value).append("\n"));
        frame.append("\n").append(body != null ? body : "").append("\u0000");
        return frame.toString();
    }

}

File: client/src/Frame.cpp
#include "Frame.h"
#include <sstream>

// Default constructor
Frame::Frame() = default;

// Parameterized constructor
Frame::Frame(const std::string& command, const std::unordered_map<std::string, std::string>& headers, const std::string& body)
    : command(command), headers(headers), body(body) {}

// Getters and setters
const std::string& Frame::getCommand() const {
    return command;
}

void Frame::setCommand(const std::string& cmd) {
    command = cmd;
}

const std::unordered_map<std::string, std::string>& Frame::getHeaders() const {
    return headers;
}

void Frame::addHeader(const std::string& key, const std::string& value) {
    headers[key] = value;
}

void Frame::removeHeader(const std::string& key) {
    headers.erase(key);
}

const std::string& Frame::getBody() const {
    return body;
}

void Frame::setBody(const std::string& bodyContent) {
    body = bodyContent;
}

// Serialize to string
std::string Frame::toString() const {
    std::ostringstream frameStream;
    frameStream << command << "\n";
    for (const auto& header : headers) {
        frameStream << header.first << ":" << header.second << "\n";
    }
    frameStream << "\n" << body << '\0'; // Null character at the end
    return frameStream.str();
}

// Parse a frame from string
Frame Frame::fromString(const std::string& frameString) {
    std::istringstream frameStream(frameString);
    std::string line;
    Frame frame;

    // Parse command
    if (std::getline(frameStream, line)) {
        frame.setCommand(line);
    }

    // Parse headers
    while (std::getline(frameStream, line) && !line.empty()) {
        auto delimiterPos = line.find(':');
        if (delimiterPos != std::string::npos) {
            std::string key = line.substr(0, delimiterPos);
            std::string value = line.substr(delimiterPos + 1);
            frame.addHeader(key, value);
        }
    }

    // Parse body
    std::string bodyContent;
    while (std::getline(frameStream, line)) {
        bodyContent += line + "\n";
    }
    if (!bodyContent.empty()) {
        bodyContent.pop_back(); // Remove the last newline
    }
    frame.setBody(bodyContent);

    return frame;
}
